### ISSUE TITLE
NoValueSentinel in Param is incorrectly serialised

### ISSUE THREAD

**Issue by talna-evogene** [apache/airflow#20442]
### Apache Airflow version

2.2.2 (latest released)

### What happened

Hi
We upgraded to version 2.2.2 and starting getting errors in the UI when we have DAG params that are numeric (tried with types 'number', 'integer' and 'float' and got the same error.

```
Python version: 3.7.12
Airflow version: 2.2.2
Node: airflow2-webserver-6df5fc45bd-5vdhs
-------------------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/models/param.py", line 61, in __init__
    jsonschema.validate(self.value, self.schema, format_checker=FormatChecker())
  File "/home/airflow/.local/lib/python3.7/site-packages/jsonschema/validators.py", line 934, in validate
    raise error
jsonschema.exceptions.ValidationError: 'NoValueSentinel' is not of type 'number'

Failed validating 'type' in schema:
    {'type': 'number'}

On instance:
    'NoValueSentinel'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.7/site-packages/flask/app.py", line 2447, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/airflow/.local/lib/python3.7/site-packages/flask/app.py", line 1952, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/airflow/.local/lib/python3.7/site-packages/flask/app.py", line 1821, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/airflow/.local/lib/python3.7/site-packages/flask/_compat.py", line 39, in reraise
    raise value
  File "/home/airflow/.local/lib/python3.7/site-packages/flask/app.py", line 1950, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/airflow/.local/lib/python3.7/site-packages/flask/app.py", line 1936, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/www/auth.py", line 51, in decorated
    return func(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/www/decorators.py", line 109, in view_func
    return f(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/www/decorators.py", line 72, in wrapper
    return f(*args, **kwargs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/utils/session.py", line 70, in wrapper
    return func(*args, session=session, **kwargs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/www/views.py", line 2426, in graph
    dag = current_app.dag_bag.get_dag(dag_id)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/utils/session.py", line 70, in wrapper
    return func(*args, session=session, **kwargs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/models/dagbag.py", line 186, in get_dag
    self._add_dag_from_db(dag_id=dag_id, session=session)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/models/dagbag.py", line 261, in _add_dag_from_db
    dag = row.dag
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/models/serialized_dag.py", line 180, in dag
    dag = SerializedDAG.from_dict(self.data)  # type: Any
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/serialization/serialized_objects.py", line 947, in from_dict
    return cls.deserialize_dag(serialized_obj['dag'])
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/serialization/serialized_objects.py", line 861, in deserialize_dag
    v = {task["task_id"]: SerializedBaseOperator.deserialize_operator(task) for task in v}
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/serialization/serialized_objects.py", line 861, in <dictcomp>
    v = {task["task_id"]: SerializedBaseOperator.deserialize_operator(task) for task in v}
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/serialization/serialized_objects.py", line 641, in deserialize_operator
    v = cls._deserialize_params_dict(v)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/serialization/serialized_objects.py", line 459, in _deserialize_params_dict
    op_params[k] = cls._deserialize_param(v)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/serialization/serialized_objects.py", line 439, in _deserialize_param
    return class_(**kwargs)
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/models/param.py", line 63, in __init__
    raise ValueError(err)
ValueError: 'NoValueSentinel' is not of type 'number'

Failed validating 'type' in schema:
    {'type': 'number'}

On instance:
    'NoValueSentinel'`
```
### What you expected to happen

shouldn't cause a serialization error. 

### How to reproduce

This is a small DAG to reproduce this error. 
We use airflow on kubernetes with:
Python version: 3.7.12
Airflow version: 2.2.2

```
from airflow import DAG
from airflow.models.param import Param
from airflow.operators.dummy import DummyOperator
from datetime import datetime
params = {
    'wd': Param(type='number', description="demo numeric param")
}
default_args ={
    'depends_on_past': False,
    'start_date': datetime(2021, 6, 16)
}
with DAG ('mydag',params=params,default_args=default_args,catchup=False,schedule_interval=None) as dag:
    t1 = DummyOperator(task_id='task1')
```

### Operating System

centos

### Deployment

Official Apache Airflow Helm Chart

### Anything else

On every dag with a numeric type of param

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by potiuk** [comment]
You are likely passing the param the value "NoValueSentinel` which is something that was used in the past to determine "NoValue", But in 2.2.2 you do not need it any more. you should simply skip that parameter rather than pass "NoValueSentinel'

How do you trigger your dags with params? Which system generates the parameters? I guess it is your system so you should fix it there.

**Comment by talna-evogene** [comment]
Hi
I don't pass this value at all. It happens when I try to open the dag in the web server even before I trigger it. It happens when I press the dag's name from the home page.

**Comment by potiuk** [comment]
I see. I think this is because the DAG is serialized in the DB with the "NoValueSentinel". Is that possible that you modify the DAG manually (for example add some empty "task" to trigger reserialisation of the DAG? 

We are going to have a CLI in the future that will trigger reserialisation for such (and similar) cases: https://github.com/apache/airflow/pull/19471 but it will only come in 2.3 (`airflow reserialize`).

In the meantime if you have more DAGs like that (and if my guess is confirmed). You could trigger re-serialization via:

```
from airflow.models.serialized_dag import SerializedDagModel
from airflow.settings import Session
session = Session()
session.query(SerializedDagModel).delete()
session.commit()
```

CC: @kaxil @collinmcnulty @uranusjr  - If I am not mistaken, it seems like this is a new scenario for "accidental" serialization-change impact, showing that re-serialization is needed. I wonder if we could - in the future - have some ways to protect in case we have similar changes (automated reserialization could solve it but I think there is too strong of performance reason against it as we discussed). 

Seems hard problem to avoid those kind of incompatibilities might easily slip through. 

What do you think if instead of reserializing everything we simply Delete the serialized DAG that caused serialization issue? That could have at least self-healing potential, but we could treat Serialization Error in a special way and even display a different error to the message ("We have a aproblem this time but check back shortly")

WDYT?

**Comment by talna-evogene** [comment]
I have tried it now on a brand new DAG.
It looks like when the param type is number this error is thrown when there is no default value. Once I set a default it works ok. 
We have DAGs that worked before and since we upgraded to 2.2.2 this behaviour started.

**Comment by uranusjr** [comment]
I don’t think this is related to reserialisation (and if I’m correct, this won’t be solved by `airflow dag reserialize`), but a bug in how `Param` checks whether a default value exists vs pickling.

@talna-evogene Can you provide some concrete steps to reproduce this? It is very difficult to know exactly what went wrong in what stage from your description.

**Comment by talna-evogene** [comment]
hi
I added a small DAG in my description to reproduce it, let me know what other information is required for you to reproduce:

This is a small DAG to reproduce this error. 
We use airflow on kubernetes with:
Python version: 3.7.12
Airflow version: 2.2.2

`from airflow import DAG
from airflow.models.param import Param
from airflow.operators.dummy import DummyOperator
from datetime import datetime
params = {
    'wd': Param(type='number', description="demo numeric param")
}
default_args ={
    'depends_on_past': False,
    'start_date': datetime(2021, 6, 16)
}
with DAG ('mydag',params=params,default_args=default_args,catchup=False,schedule_interval=None) as dag:
    t1 = DummyOperator(task_id='task1')

**Comment by subkanthi** [comment]
The problem is here in this line, because the value is of type 'str' and so the `not isinstance` isnt working as expected.

```
        # If we have a value, validate it once. May raise ValueError.
        if not isinstance(value, ArgNotSet):
            try:
                jsonschema.validate(value, schema, format_checker=FormatChecker())
            except ValidationError as err:
                raise ValueError(err)
```

<img width="1247" alt="Screen Shot 2022-01-05 at 4 01 44 PM" src="https://user-images.githubusercontent.com/17434846/148288963-ff3b4950-16e3-47d6-aa2d-b3775e9fd35e.png">

**Comment by uranusjr** [comment]
Weird, looks like a DAG serialisation issue? The `ArgNotSet` instance is casted to str on serialisation, and when the DAG is deserialised it continues to be a str.

**Comment by ashlee3209** [comment]
Hi, I faced the same issue when I am using 2.2.4. I added a type validation of boolean for a parameter then when clicking the dag page, it shows error:

> Python version: 3.7.12
Airflow version: 2.2.4
Node: 3d45b1bc1955
Traceback (most recent call last):
  File "/home/airflow/.local/lib/python3.7/site-packages/airflow/models/param.py", line 62, in __init__
    jsonschema.validate(self.value, self.schema, format_checker=FormatChecker())
  File "/home/airflow/.local/lib/python3.7/site-packages/jsonschema/validators.py", line 934, in validate
    raise error
jsonschema.exceptions.ValidationError: 'NoValueSentinel' is not of type 'null', 'boolean'
Failed validating 'type' in schema:
    {'type': ['null', 'boolean']}
On instance:
    'NoValueSentinel'

any ideas?

**Comment by ephraimbuddy** [comment]
> any ideas?

Fixed in 2.3.0. We have a beta release of 2.3.0, can you test it? See https://lists.apache.org/thread/9tsks6hg2974154w5jt1jq0j9x8wlj7k
